components/api-resource-mgt/org.wso2.carbon.identity.api.resource.collection.mgt/src/main/java/org/wso2/carbon/identity/api/resource/collection/mgt/listener/ConsoleRoleListener.java

```diff
@@ -0,0 +1,256 @@
+
+    @Override
+    public void postGetPermissionListOfRole(List<Permission> permissionListOfRole, String roleId, String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        if (isConsoleRole(roleId, tenantDomain)) {
+            List<Permission> rolePermissions = getUpdatedPermissions(permissionListOfRole, tenantDomain);
+            permissionListOfRole.retainAll(rolePermissions);
+        }
+    }
+
+    @Override
+    public void postGetPermissionListOfRoles(List<String> permissions, List<String> roleIds, String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        boolean isConsoleRole = false;
+        for (String roleId : roleIds) {
+            if (isConsoleRole(roleId, tenantDomain)) {
+                isConsoleRole = true;
+                break;
+            }
+        }
+        if (isConsoleRole) {
+            List<Permission> resolvedRolePermissions = new ArrayList<>();
+            List<Permission> systemPermissions = getSystemPermission(tenantDomain);
+            permissions.forEach(permission -> {Optional<Permission> newPermission = systemPermissions.stream()
+                        .filter(permission1 -> permission1.getName().equals(permission)).findFirst();
+                newPermission.ifPresent(resolvedRolePermissions::add);
+            });
+            List<Permission> rolePermissions = getUpdatedPermissions(resolvedRolePermissions, tenantDomain);
+            permissions.retainAll(rolePermissions.stream().map(Permission::getName).collect(Collectors.toList()));
+        }
+    }
+
+    @Override
+    public void preUpdatePermissionsForRole(String roleId, List<Permission> addedPermissions,
+                                            List<Permission> deletedPermissions, String audience, String audienceId,
+                                            String tenantDomain) throws IdentityRoleManagementException {
+
+        if (APPLICATION.equals(audience)) {
+            List<Permission> consoleFeaturePermissions = getConsoleFeaturePermissions(addedPermissions);
+            if (ArrayUtils.isNotEmpty(consoleFeaturePermissions.toArray())) {
+                // If console features are added to the role, then we need to we only need to persist the console
+                // permissions.
+                addedPermissions.retainAll(consoleFeaturePermissions);
+            }
+        }
+    }
+
+    private List<Permission> getUpdatedPermissions(List<Permission> rolePermissions, String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        // Fetch all system scopes to resolve permission details from permission name.
+        List<Permission> systemPermissions = getSystemPermission(tenantDomain);
+        List<APIResourceCollection> apiResourceCollections = getAPIResourceCollections(tenantDomain);
+        List<Permission> consoleFeaturePermissions = getConsoleFeaturePermissions(rolePermissions);
+        if (!consoleFeaturePermissions.isEmpty()) {
```

Comment:
What if we add scopes to an existing role (old role). Then its corresponding scopes will be saved console:abc:view, console:abc:edit. 
When getting the scopes of the role, it will have new formatted scopes and old role related scopes. But the logic might skip for processing old roles related scopes.